File: admin/src/types/permission.ts
export interface PermissionCondition {
  field: string;
  operator: string;
  value: any;
}

export interface PermissionRow {
  id: string;
  entity: string;
  action: string;
  roles: string[];
  conditions: PermissionCondition[] | string;
  created_at: string;
  updated_at: string;
}

export interface PermissionPayload {
  id?: string;
  entity: string;
  action: string;
  roles: string[];
  conditions: PermissionCondition[];
}

export function parseConditions(row: PermissionRow): PermissionCondition[] {
  if (typeof row.conditions === "string") {
    try {
      return JSON.parse(row.conditions);
    } catch {
      return [];
    }
  }
  return row.conditions ?? [];
}

export function emptyPermission(): PermissionPayload {
  return {
    entity: "",
    action: "read",
    roles: [],
    conditions: [],
  };
}

File: backend/golang/internal/metadata/permission.go
package metadata

// Permission represents a metadata-driven permission policy.
type Permission struct {
	ID         string              `json:"id,omitempty"`
	Entity     string              `json:"entity"`
	Action     string              `json:"action"`
	Roles      []string            `json:"roles"`
	Conditions []PermissionCondition `json:"conditions,omitempty"`
}

// PermissionCondition is a field-level condition for a permission policy.
type PermissionCondition struct {
	Field    string `json:"field"`
	Operator string `json:"operator"`
	Value    any    `json:"value"`
}
